sphinx/environment/__init__.py

```diff
@@ -79,6 +79,60 @@
     'text': is_translatable,
 }
 
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from typing_extensions import Literal, overload
+
+    from sphinx.domains.c import CDomain
+    from sphinx.domains.changeset import ChangeSetDomain
+    from sphinx.domains.citation import CitationDomain
+    from sphinx.domains.cpp import CPPDomain
+    from sphinx.domains.index import IndexDomain
+    from sphinx.domains.javascript import JavaScriptDomain
+    from sphinx.domains.math import MathDomain
+    from sphinx.domains.python import PythonDomain
+    from sphinx.domains.rst import ReSTDomain
+    from sphinx.domains.std import StandardDomain
+    from sphinx.ext.duration import DurationDomain
+    from sphinx.ext.todo import TodoDomain
+
+    class DomainsType(MutableMapping[str, Domain]):
+        @overload
+        def __getitem__(self, key: Literal["c"]) -> CDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["cpp"]) -> CPPDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["changeset"]) -> ChangeSetDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["citation"]) -> CitationDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["index"]) -> IndexDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["js"]) -> JavaScriptDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["math"]) -> MathDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["py"]) -> PythonDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["rst"]) -> ReSTDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["std"]) -> StandardDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["duration"]) -> DurationDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["todo"]) -> TodoDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: str) -> Domain: ...  # NoQA: E704
+        def __getitem__(self, key): raise NotImplementedError  # NoQA: E704
+        def __setitem__(self, key, value): raise NotImplementedError  # NoQA: E704
+        def __delitem__(self, key): raise NotImplementedError  # NoQA: E704
+        def __iter__(self): raise NotImplementedError  # NoQA: E704
+        def __len__(self): raise NotImplementedError  # NoQA: E704
```

Comment:
this approach doesn't seem right to me. What are you trying to achieve? you might be looking for `TypedDict`, rather than `MutableMapping`?

something like-

```python
class DomainsType(TypedDict):
    c: CDomain
    cpp: CPPDomain
    changeset: ChangeSetDomain
    ...
```

can you give me some more context here?

Reply:
on review i'm not sure there is a better solution than what you're proposing right now. There are feature requests to allow additional keys on `TypedDict`, but there are soundness issues with the proposal.

I guess that what this really highlights is that if this codebase had used strict type annotations from the start, it would probably not be implemented this way. I presume you're avoiding breaking changes here. If you *could* make breaking changes you would just promote the magic keys to concrete properties

Reply:
> on review i'm not sure there is a better solution than what you're proposing right now. There are feature requests to allow additional keys on `TypedDict`, but there are soundness issues with the proposal.

Thank you for your review, this is what I feared having tried a `Protocol` and `TypedDict` approach and having gotten nowhere.

> I guess that what this really highlights is that if this codebase had used strict type annotations from the start, it would probably not be implemented this way. 

Challenging given the first commit here was around seven years before [PEP 484](https://peps.python.org/pep-0484/)!

>I presume you're avoiding breaking changes here. If you _could_ make breaking changes you would just promote the magic keys to concrete properties

I'll consider the attributes approach for Sphinx 6.0/7.0, though it has drawbacks in that the user-registered domains will be second-class in some way.

A

Reply:
> the user-registered domains will be second-class in some way.

they *are* second-class, in that they are dynamically registered, instead of statically registered. It makes sense that these would be handled differently, as this is a pretty fundamental difference. I would probably go further and make this explicit-

```python
class DomainsType:
    c: CDomain
    cpp: CPPDomain
    changeset: ChangeSetDomain
    ...
    custom: dict[str, Domain]
```